MIGRATION.md

```diff
@@ -0,0 +1,423 @@
+
+<!-- markdownlint-disable code-block-style -->
+
+=== "URL-based State Handling"
+
+    ```python title="v2"
+    success, logs = verifier.verify_pacts(
+        './pacts/consumer-provider.json',
+        provider_states_setup_url='http://localhost:8080/_pact/provider_states'
+    )
+    ```
+
+    ```python title="v3"
+    # Option 1: URL-based (similar to v2)
+    verifier = (
+        Verifier('my-provider')
+        .add_transport(url='http://localhost:8080')
+        .state_handler(
+            'http://localhost:8080/_pact/provider_states',
+            body=True  # (1)
+        )
+        .add_source('./pacts/')
+    )
+    ```
+
+    1.  The `body` argument specifies whether to use a `POST` request and pass information in the body, or to use a `GET` request and pass information through HTTP headers. For more details, see the [`state_handler` API documentation][pact.verifier.Verifier.state_handler].
+
+=== "Functional State Handling"
+
+    ```python title="v2"
+    # Not supported
+    ```
+
+    ```python title="v3 - Function"
+        def handler(name, params=None):
+            if name == 'user exists':
+                # Set up user in database/mock
+                create_user(params.get('id', 123))
+            elif name == 'no users exist':
+                # Clear users
+                clear_users()
+
+        verifier = (
+            Verifier('my-provider')
+            .add_transport(url='http://localhost:8080')
+            .state_handler(handler)
+            .add_source('./pacts/')
+        )
+    ```
+
+    ```python title="v3 - Mapping"
+    state_handlers = {
+        'user exists': lambda name, params: create_user(params.get('id', 123)),
+        'no users exist': lambda name, params: clear_users(),
+    }
+
+    verifier = (
+        Verifier('my-provider')
+        .add_transport(url='http://localhost:8080')
+        .state_handler(state_handlers)
+        .add_source('./pacts/')
+    )
+    ```
+
+    More information on the state handler function signature can be found in the [`state_handler` API documentation][pact.verifier.Verifier.state_handler]. By default, the handlers only _set up_ the provider state. If you need to also _tear down_ the state after verification, you can use the `teardown=True` argument to enable this behaviour.
+
+    !!! warning
+
+        These functions run in the test process, so any side effects must be properly shared with the provider. If using mocking libraries, ensure the provider is started in a separate thread of the same process (using `threading.Thread` or similar), rather than a separate process (e.g., using `multiprocessing.Process` or `subprocess.Popen`).
+
+<!-- markdownlint-enable code-block-style -->
+
+#### Message Verification
+
+Message verification is now much more straightforward in v3:
```

Comment:
it might be worth noting that this closes a few gaps in the previous impl

some notes

https://github.com/YOU54F/pact-logical-replication/tree/main/message_pact_python#pact-python-messaging

might be worth mentioning in the migration guide, that message handlers are correctly mapped to descriptions, not providerStates, and v3 matching rules will be applied where previous v2 matching rules were.

Or closing off those linked issues as being resolved in V3, and referenced here.

nice way to mop up a few outstanding issues, and probably lead users towards using pp v3 if they want to use message pact (and I have seen it in use at customer sites)